android-components/components/browser/state/src/test/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddlewareTest.kt

```diff
@@ -622,4 +653,72 @@ class TranslationsMiddlewareTest {
             ),
         )
     }
+
+    @Test
+    fun `WHEN OperationRequestedAction is dispatched to fetch downloaded languages AND succeeds THEN SetDownloadedLanguagesAction is dispatched`() = runTest {
+        val code = "es"
+        val localizedDisplayName = "Spanish"
+        val isDownloaded = true
+        val size: Long = 1234
+        val language = Language(code, localizedDisplayName)
+        val languageModel = LanguageModel(language, isDownloaded, size)
+        val downloadedLanguages = mutableListOf(languageModel)
+        val languagesCallback = argumentCaptor<((List<LanguageModel>) -> Unit)>()
+
+        // Initial Action
+        val action =
+            TranslationsAction.OperationRequestedAction(
+                tabId = tab.id,
+                operation = TranslationOperation.FETCH_DOWNLOADED_LANGUAGES,
+            )
+        translationsMiddleware.invoke(context, {}, action)
+
+        // Verify results
+        verify(engine, atLeastOnce()).getTranslationsModelDownloadStates(onSuccess = languagesCallback.capture(), onError = any())
+        languagesCallback.value.invoke(downloadedLanguages)
+
+        verify(context.store, atLeastOnce()).dispatch(
+            TranslationsAction.SetDownloadedLanguagesAction(
+                downloadedLanguages = downloadedLanguages,
+            ),
+        )
+
+        waitForIdle()
+    }
+
+    @Test
+    fun `WHEN OperationRequestedAction is dispatched to fetch downloaded languages AND fails THEN TranslateExceptionAction is dispatched`() = runTest {
+        val errorCallback = argumentCaptor<((Throwable) -> Unit)>()
+        whenever(
+            engine.getTranslationsModelDownloadStates(
+                onSuccess = any(),
+                onError = errorCallback.capture(),
+            ),
+        ).thenAnswer { errorCallback.value.invoke(Throwable()) }
+
+        val action =
+            TranslationsAction.OperationRequestedAction(
+                tabId = tab.id,
+                operation = TranslationOperation.FETCH_DOWNLOADED_LANGUAGES,
+            )
+        translationsMiddleware.invoke(context, {}, action)
+        waitForIdle()
+
+        // Verify Dispatch
+        verify(store, atLeastOnce()).dispatch(
+            TranslationsAction.EngineExceptionAction(
+                error = TranslationError.CouldNotLoadLanguagesError(any()),
+            ),
+        )
+
+        verify(store, atLeastOnce()).dispatch(
+            TranslationsAction.TranslateExceptionAction(
```

Comment:
Could you take a quick look at this test? I sometimes think this style exception verifying tests are waiting properly to complete the verify and sometimes think they aren't. They've been giving me trouble overall. Curious what your experience with them has been and what you think too!

Reply:
Yea, I notice it's not checking that the `translationError` matches

Reply:
I'll see if I can find out more too! The errors especially seem to not validate consistently.